paddle/fluid/operators/gather_scatter_kernel.cc

```diff
@@ -0,0 +1,203 @@
+/* Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include "paddle/fluid/operators/gather_scatter_kernel.h"
+namespace paddle {
+namespace operators {
+
+using Tensor = framework::Tensor;
+
+class TensorAssign {
+ public:
+  template <typename tensor_t>
+  void operator()(tensor_t* self_data, tensor_t* src_data) const {
+    *self_data = *src_data;
+    VLOG(3) << "self_data assigned:" << *self_data;
+  }
+};
+static TensorAssign tensor_assign;
+
+class ReduceAdd {
+ public:
+  template <typename tensor_t>
+  void operator()(tensor_t* self_data, tensor_t* src_data) const {
+    *self_data += *src_data;
+    VLOG(3) << "self_data added:" << *self_data;
+  }
+};
+
+static ReduceAdd reduce_add;
+
+class ReduceMultiply {
+ public:
+  template <typename tensor_t>
+  void operator()(tensor_t* self_data, tensor_t* src_data) const {
+    *self_data *= *src_data;
+  }
+
+  void operator()(bool* self_data, bool* src_data) const {
+    *self_data = *self_data && *src_data;
+  }
+};
+
+static ReduceMultiply reduce_mul;
+
+template <typename tensor_t, typename index_t = int64_t,
+          bool is_scatter_like = true>
+struct cpu_gather_scatter_functor {
```

Comment:
I think it is better to split it into two classes; one for gather and the other for scatter

Reply:
I don't think split it is better since I think scatter and gather share a lot of logic. Making one kernel function can make the code more elegant. Don't want make some redundant code.